### PR TITLE
Validate beta features only when v1 Tasks and Pipelines are defined

### PR DESCRIPTION
**Pull request by lbernick**
Currently, validation differs between api versions: when beta features are used in v1 APIs,
"enable-api-fields" must be set to "alpha" or "beta", but when beta features are used in beta APIs,
"enable-api-fields" may be set to "alpha", "beta", or "stable".

We also validate the specs of referenced Tasks or Pipelines in the TaskRun/PipelineRun reconciler.
This presents a problem when referencing a Task or Pipeline declared locally, since the Task or Pipeline may be converted into a different API version when it's stored.

This commit moves validation for beta features to apply only to Tasks and Pipelines when they are created or updated,
and not called when a Task spec or Pipeline spec is validated.

This commit will allow us to swap the storage version of our API without user-facing impact.
Separately, we plan to decouple feature versioning from API versioning, as it is a better long-term solution (https://github.com/tektoncd/pipeline/issues/6592).

This commit contains no expected functional changes, since the TaskRun and PipelineRun reconcilers do not
currently validate that"enable-api-fields" is set to "alpha" or "beta" when beta features are used in
referenced Tasks or Pipelines.

This validation will be added for v1 remote Tasks and Pipelines in a separate commit (https://github.com/tektoncd/pipeline/pull/6725).

Closes https://github.com/tektoncd/pipeline/issues/6616.
/kind misc

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- n/a Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- n/a Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- n/a Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Review by vdemeester — COMMENTED** [review]
Couldn't we use the context (`ctx`) to kno when to validate and when to not validate (as there is already information in there about if it's at creation, or update, etc..) ?

One of the shortcomings of this is, if the webhook is not ready/registered, one might be able to create something that is not valid (aka uses beta fields while the configuration is set to `stable`) and get it executed. This, however, can only happen explicitely (user removes the webhook/admission controller registration) or at installation (in a very short period of time where CRDs are there but the webhook has not fully started…).

**Comment by lbernick** [comment]
chatted offline with @vdemeester, but this PR doesn't actually change anything from the current implementation. Right now, we don't do any validation in the reconciler for beta fields. We do call `taskSpec.Validate` and `pipelineSpec.Validate` in the reconciler, but the fetched pipeline or task is treated as v1beta1, so no validation of beta features is done: https://github.com/tektoncd/pipeline/blob/dffb5f457d64368166594087d110c462f8c6eddd/pkg/reconciler/pipelinerun/pipelinerun.go#L448
We could discuss future improvements guarding against the webhook being disabled or not ready, but this PR doesn't change our current validation behavior, it just makes it possible for us to swap to v1 as our storage version without breaking v1beta1 pipelines using beta features.

We could also choose to add beta features validation for v1 remote tasks here: https://github.com/tektoncd/pipeline/blob/dffb5f457d64368166594087d110c462f8c6eddd/pkg/reconciler/pipelinerun/resources/pipelineref.go#L163
Note that this would be a change from what we're doing today.

**Comment by vdemeester** [comment]
I am a bit neutral on this, I don't really like that approach too much, but I don't hate it either, and it does unblock switching the storage version. As a side note, I don't really see a big rush to switch the storage version (as discussed during the API WG) but I also don't see too much reason to block that either.

I would love to have opinion of others @tektoncd/core-maintainers  👼🏼

**Comment by lbernick** [comment]
@vdemeester could you elaborate on your concerns with this proposal or any problems you think we might run into? We could definitely brainstorm ways to make the reconciler more resilient to issues with the webhook validation being accidentally skipped, but I don't think that should block this PR.

**Comment by lbernick** [comment]
known flake: https://github.com/tektoncd/pipeline/issues/6672
/retest

**Comment by dibyom** [comment]
hey @lbernick this is neat and generally I like the approach. I have two questions - 
1. post v1 swap, a user has a v1beta1 pipeline that contains object params that they are referencing in a v1beta1 pipelinerun (with the default values of enable-api-fields set) - will this pipeline work or fail validation? 

2. today i.e. pre v1 swap, if I reference a v1 pipeline with beta fields via remote resolution, does that trigger a failed validation as well?

**Comment by lbernick** [comment]
> hey @lbernick this is neat and generally I like the approach. I have two questions -
> 
> 1. post v1 swap, a user has a v1beta1 pipeline that contains object params that they are referencing in a v1beta1 pipelinerun (with the default values of enable-api-fields set) - will this pipeline work or fail validation?

In the reconciler, we only call `pipelineSpec.Validate`, not `pipelineRunSpec.Validate`. We do not currently validate that enable-api-fields is set to "beta" for beta features in the reconciler, and this commit means we will still not do so even after swapping to v1. Therefore, this example should work.

> 2. today i.e. pre v1 swap, if I reference a v1 pipeline with beta fields via remote resolution, does that trigger a failed validation as well?

It doesn't. This is an example of how validation for remote Pipelines differs from validation of local Pipelines. We should probably add this validation, but I don't think that should block this PR.

**Comment by lbernick** [comment]
@dibyom I've added a third commit for validating beta features in v1 remote pipelines just to show what it looks like. (Happy to either split these commits into multiple PRs or squash them.)

**Comment by lbernick** [comment]
/retest
https://github.com/tektoncd/pipeline/issues/4169

**Comment by JeromeJu** [comment]
/lgtm
/hold
so as to allow more @tektoncd/core-maintainers would have time to look at this.

**Review by dibyom — APPROVED** [review]
Code LGTM, but let's wait till the pipeline wg next week to merge to align on the discussions around alternatives to this and why we need this PR - I wrote up my understanding of the summary here: https://github.com/tektoncd/pipeline/issues/6592#issuecomment-1564694433

**Review by dibyom — APPROVED** [review]
Code LGTM, but let's wait till the pipeline wg next week to merge to align on the discussions around alternatives to this and why we need this PR - I wrote up my understanding of the summary here: https://github.com/tektoncd/pipeline/issues/6592#issuecomment-1564694433

**Review by dibyom — APPROVED** [review]
Code LGTM, but let's wait till the pipeline wg next week to merge to align on the discussions around alternatives to this and why we need this PR - I wrote up my understanding of the summary here: https://github.com/tektoncd/pipeline/issues/6592#issuecomment-1564694433

**Comment by lbernick** [comment]
/hold cancel
removing the hold just to avoid confusion-- I wouldn't want people to avoid reviewing since there's a hold on it. @JeromeJu if you want to avoid this getting merged you can approve instead of lgtm!

**Review by JeromeJu — APPROVED** [review]

**Review by JeromeJu — APPROVED** [review]

File: pkg/apis/pipeline/v1/pipeline_validation.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1/pipeline_validation.go:73]: just to make it clear for the reviewers, this PR is moving this validation from `func (ps *PipelineSpec) Validate()` to `func (p *Pipeline) Validate()` such that the validation `ValidateBetaFields` is done during `pipeline` creation time compared to `pipelineRun`.

File: pkg/apis/pipeline/v1/task_validation_test.go
Comment (Yongxuanzhang) [inline pkg/apis/pipeline/v1/task_validation_test.go:1929]: Should we rename the test name from "TestTaskSpecBetaFields" to "TestTaskBetaFields"?

Comment (lbernick) [reply pkg/apis/pipeline/v1/task_validation_test.go:1929]: Done

**Comment by lbernick** [comment]
> Code LGTM, but let's wait till the pipeline wg next week to merge to align on the discussions around alternatives to this and why we need this PR - I wrote up my understanding of the summary here: [#6592 (comment)](https://github.com/tektoncd/pipeline/issues/6592#issuecomment-1564694433)

discussion at Pipelines WG today-- no concerns re: merging this PR

**Comment by JeromeJu** [comment]
/lgtm
as discussed in the WG and the issue itself

**Comment by lbernick** [comment]
known flake: https://github.com/tektoncd/pipeline/issues/6672
This flake is marked fixed, but it was fixed only a few hours ago so maybe the fix has not been picked up by this PR yet? I'm not sure exactly how our merge automation works
/retest

**Comment by Yongxuanzhang** [comment]
> known flake: #6672 This flake is marked fixed, but it was fixed only a few hours ago so maybe the fix has not been picked up by this PR yet? I'm not sure exactly how our merge automation works /retest

If that fix works, and the PR has rebased from the main. Then we shouldn’t encounter the flaky again.

**Pull request merged**
